### PR TITLE
`GH-188` `events_mgmt` adding support for Adobe `Runtime` Actions as webhooks 

### PR DESCRIPTION
**Pull request by abhupadh**
## Description

Adding support for managing runtime webhook registrations

## Related Issue

* #188 

## Motivation and Context

To enable integration testing for runtime webhook registrations.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the [code style](CODE_STYLE.md) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 1]
```diff
@@ -61,19 +66,23 @@ public class Workspace {
...
+                    final String runtimeNamespace, final List<String> runtimeActions, Context authContext) {
     this.imsUrl = StringUtils.isEmpty(imsUrl) ? Constants.IMS_URL : imsUrl;
     this.imsOrgId = imsOrgId;
 
     this.apiKey = apiKey;
     this.consumerOrgId = consumerOrgId;
     this.projectId = projectId;
     this.workspaceId = workspaceId;
-
+    this.runtimeNamespace = runtimeNamespace;
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:83]: are these really related to the Workspace ? 
or should we move these config elsewhere  ?

File: events_test/src/main/java/com/adobe/aio/event/publish/PublishServiceTester.java [hunk 2]
```diff
@@ -68,4 +81,22 @@ public static String getEventDataNode(String eventId) {
     return "{\"" + DATA_EVENT_ID_NODE + "\" : \"" + eventId + "\"}";
   }
 
+  public static String getEventDataNodeForRuntimeWebhook(String eventId, String environment,
+      String registrationId, String runtimeAction) {
```
Comment (francoisledroff) [inline events_test/src/main/java/com/adobe/aio/event/publish/PublishServiceTester.java:85]: do we really need this `data` node change to test runtime integration ?

**Review by francoisledroff — CHANGES_REQUESTED** [review]
a few first comments

File: core/src/main/java/com/adobe/aio/util/Constants.java [hunk 3]
```diff
@@ -19,6 +19,7 @@ public class Constants {
   public static final String IMS_ORG_HEADER = "x-gw-ims-org-id";
   public static final String IMS_URL = "https://ims-na1.adobelogin.com";
   public static final String API_MANAGEMENT_URL = "https://api.adobe.io";
+  public static final String AIO_EVENTS_WEBHOOK_APP_URL = "https://webhooks.adobe.io/egauditor";
```
[the remark this answers is not in the document]
Comment (abhupadh) [reply core/src/main/java/com/adobe/aio/util/Constants.java:22]: sure

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java [hunk 4]
```diff
@@ -61,19 +66,23 @@ public class Workspace {
   private final String consumerOrgId;
   private final String projectId;
   private final String workspaceId;
+  private final String runtimeNamespace;
+  private final List<String> runtimeActions;
```
Comment (francoisledroff) [inline core/src/main/java/com/adobe/aio/workspace/Workspace.java:70]: could we remove `runtimeActions` from this workspace very static config
as it's will be created/deleted dynamically ?

Comment (francoisledroff) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java:70]: it's not used anyway in this open source sdk, is it ?

File: events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+public class DataNodeUtil {
+
+  private static final String DATA_EVENT_ID_NODE = "data_event_id";
+
+  private DataNodeUtil() {
+    throw new IllegalStateException("Should not to be instantiated");
+  }
+
+  public static String getEventDataNode(String eventId) {
```
Comment (francoisledroff) [inline events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java:22]: wdyt moving this method within `PublishServiceIntegrationTest` ?

**Review by francoisledroff — CHANGES_REQUESTED** [review]
a few requests:

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 4 above)
Comment (abhupadh) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java:70]: correct, only `namespace` will be required. I will remove it

File: events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java (hunk 5 above)
Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java:22]: ok

Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java:22]: it is getting used in the `JournalServiceIntegrationTest` also.

Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/util/DataNodeUtil.java:22]: Let's keep it here only.

File: core/src/main/java/com/adobe/aio/workspace/Workspace.java (hunk 4 above)
Comment (abhupadh) [reply core/src/main/java/com/adobe/aio/workspace/Workspace.java]: done

**Review by francoisledroff — DISMISSED** [review]
I'll propose some cleanup in my own PR

File: events_mgmt/src/main/java/com/adobe/aio/event/management/model/RegistrationUpdateModel.java [hunk 6]
```diff
@@ -58,11 +61,16 @@ public class RegistrationUpdateModel {
     this.webhookUrl = webhookUrl;
     this.eventsOfInterestInputModels = eventsOfInterestInputModels;
     if (DeliveryType.fromFriendlyName(deliveryType).isWebhookDelivery() && StringUtils.isEmpty(webhookUrl)) {
-      throw new IllegalArgumentException(
-                      "Registration is a webhook registration, but missing a webhook url");
+      if (StringUtils.isNotEmpty(runtimeAction)) {
+        this.deliveryType = DeliveryType.WEBHOOK.getFriendlyName();
+      } else {
+        throw new IllegalArgumentException(
+            "Registration is a webhook registration, but missing a webhook url");
```
Comment (nicdard) [inline events_mgmt/src/main/java/com/adobe/aio/event/management/model/RegistrationUpdateModel.java:68]:
```suggestion
            "Registration is a webhook registration, but missing a webhook url or a runtime action");
```

File: events_mgmt/src/main/java/com/adobe/aio/event/management/model/RegistrationUpdateModel.java [hunk 7]
```diff
@@ -58,11 +61,16 @@ public class RegistrationUpdateModel {
     this.webhookUrl = webhookUrl;
     this.eventsOfInterestInputModels = eventsOfInterestInputModels;
     if (DeliveryType.fromFriendlyName(deliveryType).isWebhookDelivery() && StringUtils.isEmpty(webhookUrl)) {
-      throw new IllegalArgumentException(
-                      "Registration is a webhook registration, but missing a webhook url");
+      if (StringUtils.isNotEmpty(runtimeAction)) {
```
Comment (nicdard) [inline events_mgmt/src/main/java/com/adobe/aio/event/management/model/RegistrationUpdateModel.java:64]: Can we move this condition inside the enclosing `if` by adding `&& StringUtils.isEmpty(runtimeAction)`?

File: events_test/src/main/java/com/adobe/aio/event/management/RegistrationServiceTester.java [hunk 8]
```diff
@@ -61,19 +64,40 @@ public Registration createJournalRegistration(String registrationName,
...
+    String registrationId = registrationCreated.getRegistrationId();
     assertNotNull(registrationId);
-    assertEquals(registrationInputModel.getDescription(), registratinCreated.getDescription());
-    assertEquals(registrationInputModel.getName(), registratinCreated.getName());
-    assertEquals(registrationInputModel.getDeliveryType(), registratinCreated.getDeliveryType());
+    assertEquals(registrationInputModel.getDescription(), registrationCreated.getDescription());
+    assertEquals(registrationInputModel.getName(), registrationCreated.getName());
+    assertEquals(registrationInputModel.getDeliveryType(), registrationCreated.getDeliveryType());
+
+    if (!registrationInputModel.getDeliveryType().equals(DELIVERY_TYPE_JOURNAL)) {
```
Comment (nicdard) [inline events_test/src/main/java/com/adobe/aio/event/management/RegistrationServiceTester.java:91]: Shouldn't we check for:
```suggestion
    if (registrationInputModel.getDeliveryType().equals(DELIVERY_TYPE_WEBHOOK)) {
```

Comment (francoisledroff) [reply events_test/src/main/java/com/adobe/aio/event/management/RegistrationServiceTester.java:91]: good catch ! it the registration uses runtime, it cannot be a journal delivery

File: events_mgmt/src/main/java/com/adobe/aio/event/management/model/Event.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Objects;
+
+/**
+ * Event POJO that is fetched from the AIO Event Test Webhook App (a.k.a eg-auditor) endpoint
+ */
+@JsonInclude(Include.NON_NULL)
+public class Event {
```
Comment (francoisledroff) [inline events_mgmt/src/main/java/com/adobe/aio/event/management/model/Event.java:25]: this Event model is not belonging to this sdk
this is only for internal testing

**Comment by francoisledroff** [comment]
I pushed some cleanup against your branch see https://github.com/adobe/aio-lib-java/pull/189

**Review by nicdard — APPROVED** [review]
LGTM

**Review by nicdard — APPROVED** [review]

**Review by francoisledroff — APPROVED** [review]
